### PR TITLE
[custom channels 2/5]: Extract PART2 from mega staging branch

### PR DESCRIPTION
**Pull request by guggero**
Extracts part 2 from https://github.com/lightningnetwork/lnd/pull/8960.

### CODE REVIEW

**Comment by guggero** [comment]
I addressed all TODOs, this is ready for full second round review.

File: aliasmgr/aliasmgr.go [hunk 1]
```diff
@@ -11,6 +11,11 @@ import (
 	"github.com/lightningnetwork/lnd/lnwire"
 )
 
+// UpdateLinkAliases is a function type for a function that locates the active
+// link that matches the given shortID and triggers an update based on the
+// latest values of the alias manager.
+type UpdateLinkAliases func(shortID lnwire.ShortChannelID) error
```
Comment (ProofOfKeags) [inline aliasmgr/aliasmgr.go:17]: 😔

File: aliasmgr/aliasmgr.go [hunk 2]
```diff
@@ -220,12 +235,22 @@ func (m *Manager) populateMaps() error {
 // AddLocalAlias adds a database mapping from the passed alias to the passed
 // base SCID. The gossip boolean marks whether or not to create a mapping
 // that the gossiper will use. It is set to false for the upgrade path where
-// the feature-bit is toggled on and there are existing channels.
+// the feature-bit is toggled on and there are existing channels. The linkUpdate
+// flag is used to signal whether this function should also trigger an update
+// on the htlcswitch scid alias maps.
 func (m *Manager) AddLocalAlias(alias, baseScid lnwire.ShortChannelID,
-	gossip bool) error {
+	gossip, linkUpdate bool) error {
```
Comment (ProofOfKeags) [inline aliasmgr/aliasmgr.go:247]: 😔😔 Are we committed to this approach?

File: aliasmgr/aliasmgr.go [hunk 3]
```diff
@@ -506,6 +506,23 @@ func (m *Manager) GetPeerAlias(chanID lnwire.ChannelID) (lnwire.ShortChannelID,
...
+
+	// haveAlias returns true if the passed alias is already assigned to a
+	// channel in the baseToSet map.
+	haveAlias := func(maybeNextAlias lnwire.ShortChannelID) bool {
+		for _, aliasList := range m.baseToSet {
+			for _, alias := range aliasList {
+				if alias == maybeNextAlias {
+					return true
+				}
+			}
```
Comment (ProofOfKeags) [inline aliasmgr/aliasmgr.go:521]: fn.Any?

File: lnrpc/marshall_utils.go [hunk 4]
```diff
@@ -207,3 +209,19 @@ func UnmarshallCoinSelectionStrategy(strategy CoinSelectionStrategy,
...
+	for base, aliases := range scidMap {
+		res = append(res, &AliasMap{
+			BaseScid: base.ToUint64(),
+			Aliases: fn.Map(func(a lnwire.ShortChannelID) uint64 {
+				return a.ToUint64()
+			}, aliases),
+		})
+	}
+
+	return res
```
Comment (ProofOfKeags) [inline lnrpc/marshall_utils.go:226]: fn.Map will clean this up a lot.

File: lnwire/lnwire_test.go [hunk 5]
```diff
@@ -1442,6 +1442,29 @@ func TestLightningWireProtocol(t *testing.T) {
...
+
+			req.CustomRecords = randCustomRecords(t, r)
+
+			// Generate some random TLV records 50% of the time.
+			if r.Int31()%2 == 0 {
+				req.ExtraData = []byte{
+					0x01, 0x03, 1, 2, 3,
+					0x02, 0x03, 4, 5, 6,
+				}
+			}
```
Comment (ProofOfKeags) [inline lnwire/lnwire_test.go:1466]: I'm somewhat surprised this didn't already exist.

File: itest/lnd_forward_interceptor_test.go [hunk 6]
```diff
@@ -344,6 +346,112 @@ func testForwardInterceptorBasic(ht *lntest.HarnessTest) {
...
+
+	targetCustomRecordsStr := fmt.Sprintf(
+		"%d:[%s]", crKey, strings.Join(asciiValues, " "),
+	)
+
+	// Wait for the log entry to appear in Carol's log.
+	carol.RPC.LogLineContains([]string{
+		"Received UpdateAddHTLC(", targetCustomRecordsStr,
+		fmt.Sprintf("amt=%d", newOutAmountMsat),
+	})
```
Comment (ProofOfKeags) [inline itest/lnd_forward_interceptor_test.go:437]: I thought I mentioned this before, but I really don't think we should be doing this type of thing to assess the correctness of our code.

File: htlcswitch/packet.go [hunk 7]
```diff
@@ -95,9 +95,13 @@ type htlcPacket struct {
...
 
 	// customRecords are user-defined records in the custom type range that
-	// were included in the payload.
+	// were included in the onion payload.
 	customRecords record.CustomSet
 
+	// inWireCustomRecords are custom type range TLVs that are included
+	// in the incoming update_add_htlc wire message.
+	inWireCustomRecords lnwire.CustomRecords
+
```
Comment (ProofOfKeags) [inline htlcswitch/packet.go:104]: Since we have two of these in the same structure, can we rename the rows to make their differences apparent in the naming?

File: lnwallet/channel.go [hunk 8]
```diff
@@ -3351,8 +3354,7 @@ func genRemoteHtlcSigJobs(keyRing *CommitmentKeyRing,
 // new commitment to the remote party. The commit diff returned contains all
 // information necessary for retransmission.
 func (lc *LightningChannel) createCommitDiff(newCommit *commitment,
-	commitSig lnwire.Sig, htlcSigs []lnwire.Sig) (*channeldb.CommitDiff,
-	error) {
+	commitSig lnwire.Sig, htlcSigs []lnwire.Sig) *channeldb.CommitDiff {
```
Comment (ProofOfKeags) [inline lnwallet/channel.go:3357]: 😁

**Review by ProofOfKeags — CHANGES_REQUESTED** [review]
I'd like to avoid log analysis as a means of testing. I'd also like to not further cement the "trigger update" pattern we are currently using for getting the switch to stay in sync with the alias manager, but perhaps that is out of scope.

**Comment by ProofOfKeags** [comment]
I did a first pass. I'll do another more detailed review next week.

File: itest/lnd_forward_interceptor_test.go (hunk 6 above)
Comment (guggero) [reply itest/lnd_forward_interceptor_test.go:437]: Yeah, I agree. This was left over as a temporary solution until we could actually assert the invoice HTLCs correctly. Removed the log line check but left in the TODO about modifying the amount (that will be fully resolved in part 4).

File: aliasmgr/aliasmgr.go (hunk 2 above)
Comment (guggero) [reply aliasmgr/aliasmgr.go:247]: That's how I interpret [this comment](https://github.com/lightningnetwork/lnd/pull/8960#discussion_r1727547252) and [your response](https://github.com/lightningnetwork/lnd/pull/8960#discussion_r1727674945), yes.
Your objections have been noted.

Comment (ProofOfKeags) [reply aliasmgr/aliasmgr.go:248]: Yeah my bad I had forgotten about the resolution on that convo. Disregard.

File: lnwallet/channel.go [hunk 9]
```diff
@@ -9087,3 +9087,19 @@ func (lc *LightningChannel) MultiSigKeys() (keychain.KeyDescriptor,
 	return lc.channelState.LocalChanCfg.MultiSigKey,
 		lc.channelState.RemoteChanCfg.MultiSigKey
 }
+
+// LocalCommitmentBlob returns the custom blob of the local commitment.
+func (lc *LightningChannel) LocalCommitmentBlob() fn.Option[tlv.Blob] {
```
Comment (ProofOfKeags) [inline lnwallet/channel.go:9092]: In these situations I always ask myself, why there isn't a remote corollary to this API?

File: routing/bandwidth.go [hunk 10]
```diff
@@ -18,7 +22,40 @@ type bandwidthHints interface {
...
+// payment should be carried by a channel based on the TLV records that may be
+// present in the `update_add_htlc` message or the channel commitment itself.
+type TlvTrafficShaper interface {
+	AuxHtlcModifier
+
+	// HandleTraffic is called in order to check if the channel identified
+	// by the provided channel ID may have external mechanisms that would
+	// allow it to carry out the payment.
+	HandleTraffic(cid lnwire.ShortChannelID,
+		fundingBlob fn.Option[tlv.Blob]) (bool, error)
```
Comment (ProofOfKeags) [inline routing/bandwidth.go:39]: I think this should perhaps be renamed: `CanHandleTraffic` or something to that effect.

File: routing/bandwidth.go [hunk 11]
```diff
@@ -89,30 +130,129 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
+	return reportedBandwidth
 }
 
 // availableChanBandwidth returns the total available bandwidth for a channel
 // and a bool indicating whether the channel hint was found. If the channel is
 // unavailable, a zero amount is returned.
 func (b *bandwidthManager) availableChanBandwidth(channelID uint64,
-	amount lnwire.MilliSatoshi) (lnwire.MilliSatoshi, bool) {
+	amount lnwire.MilliSatoshi,
+	firstHopBlob fn.Option[tlv.Blob]) (lnwire.MilliSatoshi, bool) {
```
Comment (ProofOfKeags) [inline routing/bandwidth.go:249]: I think we'll want some better terminology than `firstHopBlob` here. I've been trying to figure out what to call it but the way I think about it is that the blob represents some kind of opaque "intent", "color", or "properties" or something.

I don't mind that the type is that of `tlv.Blob` but as usual I am advocating naming variables and fields by their semantic purpose rather than their representational encoding.

So the question here is that, at it's most abstract conceptualization, what *IS* the `firstHobBlob` or *any* of the "blob" named fields in this PR?

File: routing/bandwidth.go [hunk 12]
```diff
@@ -89,30 +130,129 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
+		log.Errorf("ShortChannelID=%v: failed to get bandwidth from "+
+			"external traffic shaper: %v", cid, err)
+
+		return 0
+	}
+
+	htlcAmount := amount
+	result.htlcAmount.WhenSome(func(amount lnwire.MilliSatoshi) {
+		htlcAmount = amount
+	})
```
Comment (ProofOfKeags) [inline routing/bandwidth.go:223]: `htlcAmount := result.htlcAmount.UnwrapOr(amount)`

File: routing/bandwidth.go [hunk 13]
```diff
@@ -89,30 +130,129 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
-	// Otherwise, we'll return the current best estimate for the available
-	// bandwidth for the link.
-	return link.Bandwidth()
+	// If the external traffic shaper determined the bandwidth, we'll return
+	// that value, even if it is zero (which would mean no bandwidth is
+	// available on that channel).
+	reportedBandwidth := linkBandwidth
+	result.bandwidth.WhenSome(func(bandwidth lnwire.MilliSatoshi) {
+		reportedBandwidth = bandwidth
+	})
```
Comment (ProofOfKeags) [inline routing/bandwidth.go:239]: `reportedBandwidth := result.bandwidth.UnwrapOr(linkBandwidth)`

**Review by ProofOfKeags — COMMENTED** [review]
Left some more detailed comments this time around. I think we're close

File: htlcswitch/interfaces.go [hunk 14]
```diff
@@ -357,13 +357,17 @@ type InterceptedPacket struct {
 	// IncomingAmount is the amount of the accepted htlc.
 	IncomingAmount lnwire.MilliSatoshi
 
-	// CustomRecords are user-defined records in the custom type range that
-	// were included in the payload.
-	CustomRecords record.CustomSet
+	// InOnionCustomRecords are user-defined records in the custom type
+	// range that were included in the payload.
+	InOnionCustomRecords record.CustomSet
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply htlcswitch/interfaces.go:373]: Hmm, the stuff in `record` is actually "official", in that we're setting custom records, but these are protocols that are fully defined. While `lnwire.CustomRecords` is in place so we can properly merge official records with custom records, and not have them clobber each other when we go to encode/decode.

File: lnwallet/channel.go (hunk 9 above)
Comment (Roasbeef) [reply lnwallet/channel.go:9092]: I think in this case, we use it within the router, and we're usually interested in our local balance.

File: routing/bandwidth.go (hunk 11 above)
Comment (Roasbeef) [reply routing/bandwidth.go:249]: > what IS the firstHobBlob or any of the "blob" named fields in this PR?

When it's `tlv.Blob`, it's just an opauqe TLV blob. It might go into a wire message in which case we morph it into `lnwire.CustomRecords`. When it doesn't go into a wire message it's something passed to a higher layer for intepretation. 

In this case, `firstHopBlob` is used to inform how path finding should function. If specified here, then we may only want to use channels that can handle that blob (based on the `TrafficShaper`). 

In short, the blobs are a way for `lnd` to maintain minimal knowledge w.r.t what the higher level is doing, while still being able to enforce normal TLV invariants such as no duplicate keys or the keys are in order.

Comment (bitromortac) [reply routing/bandwidth.go:249]: Is it possible to specify different asset types in `lnwire.CustomRecords` for payment splitting? It feels like there is something in analogy to a payment and its multiple parts where the initial custom records represent the payment and the modified custom records represent the different htlcs or parts.

File: itest/lnd_forward_interceptor_test.go [hunk 15]
```diff
@@ -436,33 +440,309 @@ func testForwardInterceptorModifiedHtlc(ht *lntest.HarnessTest) {
...
+	ht.CloseChannel(alice, cpAB)
+	ht.CloseChannel(bob, cpBC)
+	ht.CloseChannel(carol, cpCD)
+}
+
+// testForwardInterceptorRestart tests that the interceptor can read any wire
+// custom records provided by the sender of a payment as part of the
+// update_add_htlc message and that those records are persisted correctly and
+// re-sent on node restart.
+func testForwardInterceptorRestart(ht *lntest.HarnessTest) {
```
Comment (bitromortac) [inline itest/lnd_forward_interceptor_test.go:556]: nice test!

**Review by bitromortac — COMMENTED** [review]
The commit structure is much nicer now, thank you :pray:, also great work with the persistence! I think it's almost ready. I have a suggestion to add the first hop custom records to the bandwidth manager, which may clean things up and hide them from pathfinding.

File: routing/bandwidth.go (hunk 11 above)
Comment (ProofOfKeags) [reply routing/bandwidth.go:249]: I understand the opacity. I also understand we don't want to make the opaque transparent. But there is some abstraction that covers all the blobs. After all, "blob" is a description of *representation*, not a description of *purpose*

File: lnrpc/routerrpc/router_server.go [hunk 16]
```diff
@@ -1531,6 +1544,138 @@ func (s *Server) HtlcInterceptor(stream Router_HtlcInterceptorServer) error {
...
+
+			// If not, let's add it to the alias manager now.
+			aliasScid := lnwire.NewShortChanIDFromInt(rpcAlias)
+
+			// But we only add it, if it's a valid alias, as defined
+			// by the BOLT spec.
+			if !aliasmgr.IsAlias(aliasScid) {
+				return nil, fmt.Errorf("SCID alias %v is not "+
+					"a valid alias", aliasScid)
+			}
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnrpc/routerrpc/router_server.go:1648]: Makes sense, yes. Even though we just query the alias manager once for all the aliases and cache the result, from the flow of things this should come first.

File: lnrpc/routerrpc/router_server.go [hunk 17]
```diff
@@ -1531,6 +1544,138 @@ func (s *Server) HtlcInterceptor(stream Router_HtlcInterceptorServer) error {
...
+				}
+
+				if exists {
+					return true, nil
+				}
+			}
+		}
+
+		return false, nil
+	}
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnrpc/routerrpc/router_server.go:1616]: IMO not in this case. Since this allows us to operate on the cached list of existing aliases. With a method on the alias manager we'd query more often, holding the lock more often.

File: itest/lnd_routing_test.go [hunk 18]
```diff
@@ -746,6 +747,194 @@ func testInvoiceRoutingHints(ht *lntest.HarnessTest) {
...
+	// Add the alias to Carol.
+	carol.RPC.XAddLocalChanAliases(&routerrpc.AddAliasesRequest{
+		AliasMaps: ephemeralAliasMap,
+	})
+
+	// We shouldn't be able to add the same alias again.
+	err = carol.RPC.XAddLocalChanAliasesErr(&routerrpc.AddAliasesRequest{
+		AliasMaps: ephemeralAliasMap,
+	})
+	require.ErrorContains(ht, err, routerrpc.ErrAliasAlreadyExists.Error())
```
[the remark this answers is not in the document]
Comment (guggero) [reply itest/lnd_routing_test.go:822]: We can't do that across RPC boundaries, unfortunately. So substring matching against the error is the best we can do. Same for the above.

File: lnwire/custom_records.go [hunk 19]
```diff
@@ -150,27 +145,104 @@ func (c CustomRecords) RecordProducers() []tlv.RecordProducer {
...
+
+	return tlv.MapToRecords(tlvMapGeneric)
+}
+
+// RecordsAsProducers converts a slice of records into a slice of record.
+func RecordsAsProducers(records []tlv.Record) []tlv.RecordProducer {
+	return fn.Map(func(record tlv.Record) tlv.RecordProducer {
+		return &recordProducer{record}
+	}, records)
+}
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnwire/custom_records.go:201]: Ah yes, you're right... Though the `Record()` method on `tlv.Record` has pointer access, and we're using non-pointer slices of records. So we still have to do:
```
	return fn.Map(func(record tlv.Record) tlv.RecordProducer {
		return &record
	}, records)
```

I guess a nice helper function for the `fn` package would be `Ptr` which returns the pointer of a type.
With that it could just be `fn.Map(fn.Ptr, records)`.

File: lnwire/extra_bytes.go [hunk 20]
```diff
@@ -120,3 +133,110 @@ func EncodeMessageExtraData(extraData *ExtraOpaqueData,
 	// are all properly sorted.
 	return extraData.PackRecords(recordProducers...)
 }
+
+// ParseAndExtractCustomRecords parses the given extra data into the passed-in
+// records, then returns any remaining records split into custom records and
+// extra data.
+func ParseAndExtractCustomRecords(allExtraData ExtraOpaqueData,
+	knownRecords ...tlv.RecordProducer) (CustomRecords,
+	map[tlv.Type]struct{}, ExtraOpaqueData, error) {
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnwire/extra_bytes.go:142]: Ah yes, I keep forgetting that type exists...

File: lnwire/extra_bytes.go [hunk 21]
```diff
@@ -120,3 +133,110 @@ func EncodeMessageExtraData(extraData *ExtraOpaqueData,
...
+
+	extraDataTlvMap, err := allExtraData.ExtractRecords(knownRecords...)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Remove the known and now extracted records from the leftover extra
+	// data map.
+	parsedKnownRecords := make(map[tlv.Type]struct{}, len(knownRecords))
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnwire/extra_bytes.go:151]: Actually, I think we'll want the pre-allocation here. So leaving it as `make(fn.Set[tlv.Type], len(knownRecords))`.

File: routing/bandwidth.go (hunk 11 above)
Comment (bitromortac) [reply routing/bandwidth.go:249]: Maybe something like `BandwidthTypeSelector` comes close? In the asset sense we select the asset type and the amount, but it is not clear when used in a more generic way with other custom channel implementations if that's still accurate.

File: htlcswitch/interfaces.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (guggero) [reply htlcswitch/interfaces.go:373]: Hmm, actually, I just noticed that these should all be `lnwire.CustomRecords`... That's the type we'll actually want to use, since we added all those additional validation methods to them.

File: routing/bandwidth.go (hunk 11 above)
Comment (guggero) [reply routing/bandwidth.go:249]: But that's the thing: `lnd` doesn't even know what the purpose of this thing is. It's completely unknown. So IMO "blob" is both a representation _and_ the superset of what we know about it in this context.
Also, we've used this terminology for the data we store everywhere now, IMO it would be more confusing to start changing it now for just this purpose.

File: lnrpc/routerrpc/router.proto [hunk 22]
```diff
@@ -176,6 +176,24 @@ service Router {
     */
     rpc UpdateChanStatus (UpdateChanStatusRequest)
         returns (UpdateChanStatusResponse);
+
+    /*
+    XAddLocalChanAliases is an experimental API that creates a set of new
+    channel SCID alias mappings. The list of successfully created mappings is
+    returned. This is only a locally stored alias, and will not be communicated
+    to the channel peer via any message. Therefore, routing over such an alias
+    will only work if the peer also calls this same RPC on their end. If an
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnrpc/routerrpc/router.proto:185]: You are correct, for _just_ the routing part it's enough if only one side sets this. But for the custom channel functionality (especially the RFQ part where the custom data within an HTLC might reference such an SCID alias) the alias needs to be known to both parties to fully work.

We added the X because of the requirement of both nodes needing to call this individually and it not being negotiated over a custom p2p message between the peers. I think it's still the goal to eventually (not this PR) make it possible to communicate this over the wire, that's when we're going to remove the X.

File: lnrpc/routerrpc/router_server.go [hunk 23]
```diff
@@ -1531,6 +1544,138 @@ func (s *Server) HtlcInterceptor(stream Router_HtlcInterceptorServer) error {
...
+
+			// But we only add it, if it's a valid alias, as defined
+			// by the BOLT spec.
+			if !aliasmgr.IsAlias(aliasScid) {
+				return nil, fmt.Errorf("SCID alias %v is not "+
+					"a valid alias", aliasScid)
+			}
+
+			err = s.cfg.AliasMgr.AddLocalAlias(
+				aliasScid, baseScid, false, true,
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnrpc/routerrpc/router_server.go:1651]: That was pre-existing and I'm not 100% I understand it either.
The description reads:
```
The gossip boolean marks whether or not to create a mapping
that the gossiper will use.
```

So I assume it's only set to `true` if we'd ever expect gossip messages referencing such an alias? Not sure when that would be the case and when not.

File: lnrpc/routerrpc/router_server.go [hunk 24]
```diff
@@ -1531,6 +1544,138 @@ func (s *Server) HtlcInterceptor(stream Router_HtlcInterceptorServer) error {
...
+					"alias, base_scid=%v, alias_scid=%v",
+					baseScid, aliasScid)
+			}
+
+			scidMap[baseScid] = append(scidMap[baseScid], aliasScid)
+		}
+	}
+
+	return &AddAliasesResponse{
+		AliasMaps: lnrpc.MarshalAliasMap(scidMap),
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnrpc/routerrpc/router_server.go:1664]: Yeah, you're right. This stemmed from a version where we skipped already existing aliases. But since we now return an error this makes more sense.

File: routing/payment_session.go [hunk 25]
```diff
@@ -139,7 +139,9 @@ type PaymentSession interface {
 	// A noRouteError is returned if a non-critical error is encountered
 	// during path finding.
 	RequestRoute(maxAmt, feeLimit lnwire.MilliSatoshi,
-		activeShards, height uint32) (*route.Route, error)
+		activeShards, height uint32,
+		firstHopCustomRecords lnwire.CustomRecords) (*route.Route,
```
[the remark this answers is not in the document]
Comment (guggero) [reply routing/payment_session.go:143]: Yeah, would make sense. I'm trying not to change the commit structure too much at this point though, so I'm going to skip this.

File: routing/router.go [hunk 26]
```diff
@@ -1277,6 +1286,11 @@ func (r *ChannelRouter) sendPayment(ctx context.Context,
 		return [32]byte{}, nil, err
 	}
 
+	// Validate the custom records before we attempt to send the payment.
```
[the remark this answers is not in the document]
Comment (guggero) [reply routing/router.go:1296]: Yeah, would make sense. I'm trying not to change the commit structure too much at this point though, so I'm going to skip this.

File: routing/bandwidth.go [hunk 27]
```diff
@@ -89,30 +130,129 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
+		// reported by the external traffic shaper. If the external
+		// traffic shaper is not handling the channel, this value will
+		// be fn.None
+		bandwidth fn.Option[lnwire.MilliSatoshi]
+
+		// htlcAmount is the amount we're going to use to check if we
+		// can add another HTLC to the channel. If the external traffic
+		// shaper is handling the channel, we'll use 0 to just sanity
+		// check the number of HTLCs on the channel, since we don't know
+		// the actual HTLC amount that will be sent.
```
[the remark this answers is not in the document]
Comment (guggero) [reply routing/bandwidth.go:147]: We don't know the HTLC amount yet because the external traffic shaper might want to send some of the value of this HTLC in a different form (e.g. assets). So the actual BTC amount that goes on-chain might be just above dust, but the actually transported "worth" might be millions of satoshis.

File: htlcswitch/mock.go [hunk 28]
```diff
@@ -920,6 +921,10 @@ func (f *mockChannelLink) ChannelPoint() wire.OutPoint {
 	return wire.OutPoint{}
 }
 
+func (f *mockChannelLink) ChannelCustomBlob() fn.Option[tlv.Blob] {
```
[the remark this answers is not in the document]
Comment (guggero) [reply htlcswitch/mock.go:924]: Hmm, not sure I understand what you mean? I don't see any commit where this is removed, neither in this PR nor the rest of the saga?

File: routing/payment_lifecycle.go [hunk 29]
```diff
@@ -722,6 +733,75 @@ func (p *paymentLifecycle) sendAttempt(
...
+
+		amount fn.Option[lnwire.MilliSatoshi]
+	}
+
+	// If a hook exists that may affect our outgoing message, we call it now
+	// and apply its side effects to the UpdateAddHTLC message.
+	result, err := fn.MapOptionZ(
+		p.router.cfg.TrafficShaper,
+		func(ts TlvTrafficShaper) fn.Result[extraDataRequest] {
+			newAmt, newRecords, err := ts.ProduceHtlcExtraData(
```
[the remark this answers is not in the document]
Comment (guggero) [reply routing/payment_lifecycle.go:765]: Yeah, that's on purpose. It might be that the custom routing information negotiated with the peer (RFQ) has expired in the meantime. So we should always call into this method so we can error out if the payment cannot succeed anymore.
Also, the amount returned here is kind of static (always just above dust), so it shouldn't change between calls. Or did you mean that it should be persisted as well?

File: lnwire/extra_bytes.go [hunk 30]
```diff
@@ -120,3 +133,110 @@ func EncodeMessageExtraData(extraData *ExtraOpaqueData,
...
+	// Construct a slice of all the records that we should include in the
+	// message extra data field. We will start by including any records from
+	// the extra data field.
+	mergedRecords, err := extraData.RecordProducers()
+	if err != nil {
+		return nil, err
+	}
+
+	// Merge the known and extra data records.
+	mergedRecords = append(mergedRecords, knownRecords...)
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnwire/extra_bytes.go:224]: Yeah, good point. Added that.

File: routing/bandwidth.go [hunk 31]
```diff
@@ -29,8 +66,9 @@ type getLinkQuery func(lnwire.ShortChannelID) (
 // uses the link lookup provided to query the link for our latest local channel
 // balances.
 type bandwidthManager struct {
-	getLink    getLinkQuery
-	localChans map[lnwire.ShortChannelID]struct{}
+	getLink       getLinkQuery
+	localChans    map[lnwire.ShortChannelID]struct{}
+	trafficShaper fn.Option[TlvTrafficShaper]
```
[the remark this answers is not in the document]
Comment (guggero) [reply routing/bandwidth.go:71]: Hmm, yeah, that makes sense. Done.

File: lnrpc/routerrpc/router.proto [hunk 32]
```diff
@@ -981,6 +1008,9 @@ message ForwardHtlcInterceptRequest {
     // The block height at which this htlc will be auto-failed to prevent the
     // channel from force-closing.
     int32 auto_fail_height = 10;
+
+    // The custom records of the peer's incoming p2p wire message.
+    map<uint64, bytes> in_wire_custom_records = 11;
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnrpc/routerrpc/router.proto:1013]: No, unfortunately not. Because the `HtlcInterceptor` only fires for _forwarded_ packages. But Carol is the destination hop here. But we do test something similar in `testForwardInterceptorWireRecords` where we have a 4 node network.

File: htlcswitch/interfaces.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply htlcswitch/interfaces.go:373]: Case in point :). What's the difference? Does one of those types refine the other one (is it a subset/more restrictive)?

File: routing/bandwidth.go [hunk 33]
```diff
@@ -40,11 +82,14 @@ type bandwidthManager struct {
...
 func newBandwidthManager(graph Graph, sourceNode route.Vertex,
-	linkQuery getLinkQuery) (*bandwidthManager, error) {
+	linkQuery getLinkQuery, firstHopBlob fn.Option[tlv.Blob],
+	trafficShaper fn.Option[TlvTrafficShaper]) (*bandwidthManager, error) {
 
 	manager := &bandwidthManager{
-		getLink:    linkQuery,
-		localChans: make(map[lnwire.ShortChannelID]struct{}),
+		getLink:       linkQuery,
+		localChans:    make(map[lnwire.ShortChannelID]struct{}),
```
Comment (ProofOfKeags) [inline routing/bandwidth.go:90]: `fn.NewSet`?

**Review by ProofOfKeags — APPROVED** [review]

File: htlcswitch/mock.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (bitromortac) [reply htlcswitch/mock.go:924]: Sorry, confused myself :), `ChannelCustomBlob` doesn't seem to be used.

**Review by bitromortac — APPROVED** [review]
LGTM :rocket:

File: routing/bandwidth.go (hunk 33 above)
Comment (guggero) [reply routing/bandwidth.go:90]: This is pre-exising and only in the diff because of a formatting/alignment change. So I'd rather not add more refactor to the PR.
